### PR TITLE
pathplanner: Don't follow invalid waypoints

### PR DESCRIPTION
**Pull request by mlyle**
I've run this through several cycles and it looks happy / like it eliminates the flyaway bug.

### CODE REVIEW

**Comment by peabody124** [comment]
1. confirmed with next that having an invalid waypoint triggers a PH and the error state for path planner (so at least it isn't a flyaway potential).

**Comment by peabody124** [comment]
ok and with this setting a middle waypoint to invalid, it still stops at the last valid waypoint but there is no longer an error condition (because we are explicitly handling this and not letting it slip) so the firmware behavior changes seem good.

File: flight/Modules/PathPlanner/pathplanner.c [hunk 1]
```diff
@@ -286,6 +284,30 @@ static void holdLastPosition()
...
+	// Get the activated waypoint
+	WaypointData waypoint;
+	WaypointInstGet(idx, &waypoint);
+
+	// Perhaps should fully validate here..
+	if (waypoint.Mode == WAYPOINT_MODE_INVALID) {
+		return false;
+	}
+
+	if (waypoint.Mode == WAYPOINT_MODE_STOP) {
```
Comment (peabody124) [inline flight/Modules/PathPlanner/pathplanner.c]: Does this stop mode have any meaning? Should we drop it too?

Comment (mlyle) [inline flight/Modules/PathPlanner/pathplanner.c]: I think we should but wasn't sure that was the minimal fix.

**Comment by mlyle** [comment]
Note it still is potentially flyaway because of the GCS half of the problem..

File: flight/Modules/PathPlanner/pathplanner.c (hunk 1 above)
Comment (peabody124) [inline flight/Modules/PathPlanner/pathplanner.c]: Good point. Maybe add this to #1669

File: ground/gcs/src/plugins/pathplanner/modeluavoproxy.cpp [hunk 2]
```diff
@@ -69,16 +69,25 @@ bool ModelUavoProxy::modelToObjects()
...
+    int x;
+
+    // Figure out the total amount of work to do.
+    int progressMax = myModel->rowCount();
+
+    if (instances > progressMax) {
+        progressMax = instances;
+    }
+
+    for(x=0;x<myModel->rowCount();++x)
```
Comment (peabody124) [inline ground/gcs/src/plugins/pathplanner/modeluavoproxy.cpp]: spacing around operators (although i see this was code from L72)

File: ground/gcs/src/plugins/pathplanner/modeluavoproxy.cpp [hunk 3]
```diff
@@ -124,6 +132,25 @@ bool ModelUavoProxy::modelToObjects()
             m_eventloop.exec();
         }
     }
+
+    for (; x < instances; x++) {
+        Waypoint *wp = Waypoint::GetInstance(objManager, x);
+                // shadows wp above
```
Comment (peabody124) [inline ground/gcs/src/plugins/pathplanner/modeluavoproxy.cpp]: what does this mean?

File: ground/gcs/src/plugins/pathplanner/modeluavoproxy.cpp [hunk 4]
```diff
@@ -124,6 +132,25 @@ bool ModelUavoProxy::modelToObjects()
             m_eventloop.exec();
         }
     }
+
+    for (; x < instances; x++) {
```
Comment (peabody124) [inline ground/gcs/src/plugins/pathplanner/modeluavoproxy.cpp]: could you add a comment that this runs over any pre-existing waypoints the model doesn't correspond to (or not - author's preference :D)

**Comment by peabody124** [comment]
Ok. GCS side changes are good too and test correctly. There are few cosmetic changes that would be nice at the author's discretion, but good to merge.

File: ground/gcs/src/plugins/pathplanner/modeluavoproxy.cpp (hunk 3 above)
Comment (mlyle) [inline ground/gcs/src/plugins/pathplanner/modeluavoproxy.cpp]: It's a for loop without initializing x first.  e.g. continue counting up from where we were.

**Comment by mlyle** [comment]
I'll touch this up in a few minutes.

**Comment by mlyle** [comment]
OK for merge.

**Pull request merged**